ldap/servers/slapd/upgrade.c

```diff
@@ -330,6 +408,182 @@ upgrade_remove_subtree_rename(void)
     return UPGRADE_SUCCESS;
 }
 
+/*
+ * Upgrade the base config of the PAM PTA plugin.
+ *
+ * Check the plugins base DN for any PTA configuration attributes. If any are found
+ * they are migrated to a child entry under the plugins base DN, and the original attrs
+ * are deleted from the base entry.
+ */
+static upgrade_status
+upgrade_pam_pta_default_config(void)
+{
+    const char *parent_dn_str = "cn=PAM Pass Through Auth,cn=plugins,cn=config";
+    char *child_cn_str = "default";
+    Slapi_Attr **present_attrs = NULL;
+    Slapi_PBlock *pb = NULL;
+    int num_attrs = 0;
+    int rc = 0;
+    int result = LDAP_SUCCESS;
+    size_t i = 0;
+
+    /* Check base entry exists */
+    Slapi_Entry *base_entry = NULL;
+    Slapi_DN *parent_sdn = slapi_sdn_new_dn_byval(parent_dn_str);
+    rc = slapi_search_internal_get_entry(parent_sdn, NULL, &base_entry, plugin_get_default_component_id());
+    if (rc != LDAP_SUCCESS || !base_entry) {
+        slapi_log_error(SLAPI_LOG_ERR, "upgrade_pam_pta_default_config",
+                        "Base entry not found: %s\n", parent_dn_str);
+
+        slapi_sdn_free(&parent_sdn);
+        return UPGRADE_FAILURE;
+    }
+
+    /* Check if the base entry contains config attrs */
+    pb = slapi_pblock_new();
+    rc = get_pam_default_attrs(parent_dn_str, &present_attrs, &num_attrs);
```

Comment:
Why not using the already retrieved parent_dn_str (base_entry) it prevents a second search

Reply:
Good question :)

Refactored to remove the second internal search and extract the config attrs from the base entry directly. Validated with PTA plugin using  Win AD server as auth backend, ran asan, no leaks.